signing/src/main/java/tech/pegasys/web3signer/signing/config/SignerLoader.java

```diff
@@ -44,191 +56,698 @@
- * The SignerLoader loads the metadata files and converts them to ArtifactSigners. This class keeps
- * track of the metadata files and ArtifactSigners that have been read and only reads them again if
- * they have been modified. It also removes the cached ArtifactSigners if the metadata file has been
- * removed.
+ * The SignerLoader loads metadata files and converts them to ArtifactSigners using a smart caching
+ * mechanism. This class maintains a cache of loaded signers and automatically handles file
+ * additions and deletions. It leverages Java virtual threads for efficient parallel processing of
+ * large file sets.
+ *
+ * <p>Processing strategies:
+ *
+ * <ul>
+ *   <li>Sequential: For small batches (below sequential threshold, default 100 files)
+ *   <li>Parallel with virtual threads: For larger batches, processed in configurable batch sizes
+ * </ul>
+ *
+ * <p>This class implements {@link Closeable} for proper resource management. The {@link #close()}
+ * method is idempotent as required by the Closeable contract.
+ *
+ * <p>This class is thread-safe and uses volatile references with immutable maps for lock-free
+ * reads. All write operations are atomic through copy-on-write semantics.
  */
-public class SignerLoader {
-
+public class SignerLoader implements Closeable {
   private static final Logger LOG = LogManager.getLogger();
-  private static final long FILES_PROCESSED_TO_REPORT = 10;
   private static final Set<String> CONFIG_FILE_EXTENSIONS = Set.of("yaml", "yml");
 
-  record CachedArtifactSigners(
-      Path metadataFile, FileTime lastModifiedTime, Set<ArtifactSigner> artifactSigners) {}
+  private static final int DEFAULT_SEQUENTIAL_THRESHOLD = 100;
+  private static final int DEFAULT_BATCH_SIZE = 500;
+  private static final int DEFAULT_TASK_TIMEOUT_SECONDS = 60;
+
+  // Required parameters (set via builder)
+  private final Path configsDirectory;
+
+  // Optional parameters with defaults
+  private final boolean parallelProcess;
+  private final int batchSize;
+  private final int taskTimeoutSeconds;
+  private final int sequentialThreshold;
+
+  private final ExecutorService virtualThreadExecutor;
+
+  private volatile Map<String, CachedSignerData> cachedArtifactSigners = Collections.emptyMap();
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+
+  /** Holds cached signer data along with file metadata for cache invalidation. */
+  private record CachedSignerData(
+      String filePath, FileTime lastModifiedTime, Set<ArtifactSigner> signers) {}
+
+  /** Private constructor - use {@link Builder} to create instances. */
+  private SignerLoader(final Builder builder) {
+    this.configsDirectory = builder.configsDirectory.toAbsolutePath().normalize();
+    this.parallelProcess = builder.parallelProcess;
+    this.batchSize = Math.max(100, builder.batchSize);
+    this.taskTimeoutSeconds = Math.max(1, builder.taskTimeoutSeconds);
+    this.sequentialThreshold = Math.max(1, builder.sequentialThreshold);
```

Comment:
Similar to above I think the defaults can be handled just in config rather than here and the config